merqube_client_lib/secapi/client.py

```diff
@@ -0,0 +1,143 @@
+from typing import Any, Iterable, Optional
+
+from cachetools import LRUCache, TTLCache, cached, cachedmethod
+
+from merqube_client_lib.constants import DEFAULT_TTL_CACHE
+from merqube_client_lib.session import MerqubeAPISession, get_merqube_session
+from merqube_client_lib.types.secapi import (
+    AddlSecapiOptions,
+    MappingTable,
+    SecapiMetricDefinition,
+    SecAPIRecordsResponse,
+)
+
+
+class _ClientBase:
+    """
+    base class that contains validation functions
+    """
+
+    def __init__(
+        self,
+        session: Optional[MerqubeAPISession] = None,
+        token: Optional[str] = None,
+        **session_kwargs: Any,
+    ):
+        self.session = session or get_merqube_session(token=token, **session_kwargs)
+        self.type_cache = TTLCache(1, ttl=DEFAULT_TTL_CACHE)  # type: ignore
+
+    def _collection_helper(
+        self,
+        *,
+        url: str,
+        query_options: dict[str, str | Iterable[str] | None] | None = None,
+    ) -> SecAPIRecordsResponse:
+        """
+        common function to /security metrics and definitions
+        """
+        options: dict[str, str | list[str]] = {}
+
+        for qo, v in (query_options or {}).items():
+            if v is not None:
+                options[qo] = v if isinstance(v, str) else ",".join(v)
+
+        return self.session.get_collection(url=url, options=options)
+
+    @cachedmethod(operator.attrgetter("type_cache"))
+    def _validate_secapi_type(self, sec_type: str) -> None:
+        """Validate security_type"""
+        res = self.session.get_collection("/security")
+        assert sec_type in (supported_types := [x["name"] for x in res]), f"sec_type must be one of {supported_types}"
+
+    def _validate_single(self, *, sec_type: str, sec_id: str | None = None, sec_name: str | None = None) -> None:
+        """Validate the input for functions that query for a single security"""
+        self._validate_secapi_type(sec_type=sec_type)
+        assert sec_id or sec_name, "Must provide either sec_id or sec_name"
+        assert not (sec_id and sec_name), "Must provide either sec_id or sec_name, not both"
+
+    def _validate_multiple(
+        self,
+        *,
+        sec_type: str,
+        sec_names: Optional[str | Iterable[str]] = None,
+        sec_ids: Optional[str | Iterable[str]] = None,
+    ) -> None:
+        """Validate the input for functions that query for one or more securities"""
+        self._validate_secapi_type(sec_type=sec_type)
+        assert sec_ids or sec_names, "Must provide either sec_ids or sec_names"
+        assert not (sec_ids and sec_names), "Must provide either sec_ids or sec_names, not both"
+
+
+class SecAPIClient(_ClientBase):
+    """
+    Secapi client class
+    """
+
+    def __init__(
+        self,
+        session: Optional[MerqubeAPISession] = None,
+        token: Optional[str] = None,
+        **session_kwargs: Any,
+    ):
+        """
+        We use partials to preserve types, see my SO post here
```

Comment:
I'm not sure what this comment is referring to. I don't see any partials anywhere in the module

Reply:
Thank you - leftover comment (i refactored this to move the methods directly under this, rather than the partials we use internally)